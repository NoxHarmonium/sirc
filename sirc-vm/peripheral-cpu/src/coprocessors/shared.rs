use peripheral_bus::{device::BusAssertions, BusPeripheral};

use crate::registers::{ExceptionUnitRegisters, Registers, SegmentedAddress};

#[derive(Default, FromPrimitive, ToPrimitive, Debug, PartialEq, Eq)]
pub enum ExecutionPhase {
    #[default]
    InstructionFetchLow = 0x0,
    InstructionFetchHigh = 0x1,
    InstructionDecode = 0x2,
    ExecutionEffectiveAddressExecutor = 0x3,
    MemoryAccessExecutor = 0x4,
    WriteBackExecutor = 0x5,
}

pub trait Executor {
    const COPROCESSOR_ID: u8;

    fn step<'a>(
        &mut self,
        phase: &ExecutionPhase,
        cause_register_value: u16,
        registers: &'a mut Registers,
        eu_registers: &'a mut ExceptionUnitRegisters,
        bus_assertions: BusAssertions,
    ) -> BusAssertions;
}

/// Extends an 8 bit value that is signed to a 16 bit value
///
/// Most of the values we pass around in the virtual machine are represented as unsigned
/// (even if they are technically 2's compliment signed)
/// This is fine most of the time, however, if we want to cast a smaller value to a
/// larger value, if the compiler doesn't know that the bits actually represent a
/// 2's complement number, it will fill the new space with zeros which will cause
/// the number's value to change.
///
///
/// ```
/// use peripheral_cpu::coprocessors::shared::sign_extend_small_offset;
///
/// assert_eq!(sign_extend_small_offset(0xFF), 0xFFFF);
/// assert_eq!(sign_extend_small_offset(0x00), 0x0000);
/// ```
///
#[must_use]
#[allow(clippy::cast_sign_loss, clippy::cast_possible_wrap)]
pub fn sign_extend_small_offset(small_offset: u8) -> u16 {
    // 1. Convert to i8 so the compiler knows we are dealing with a signed number
    // 2. Extend it to i16, the compiler knows it is signed so it preserves the sign
    // 3. Cast it back to unsigned which we deal with
    i16::from(small_offset as i8) as u16
}

// TODO: Rename to something more generic like fetch double word
#[must_use]
pub fn fetch_instruction(mem: &mut BusPeripheral, pc: (u16, u16)) -> [u8; 4] {
    // Only the CPU knows that the address is split into two 16 bit registers
    // Any other peripheral will only see the 24 address lines
    let full_address: u32 = pc.to_full_address();

    // TODO: Alignment check?
    // TODO: Do we need to copy here?
    let [b1, b2] = u16::to_be_bytes(mem.read_address(full_address).to_owned());
    let [b3, b4] = u16::to_be_bytes(mem.read_address(full_address + 1).to_owned());
    [b1, b2, b3, b4]
}
