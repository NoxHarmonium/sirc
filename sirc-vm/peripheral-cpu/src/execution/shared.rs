use peripheral_mem::MemoryPeripheral;

use crate::{
    instructions::definitions::{ConditionFlags, Instruction, StatusRegisterUpdateSource},
    registers::Registers,
};

/**
* The instruction mapped out into components.
*
* Simulates the temporary registers the CPU would have when an instruction
* is being decoded.
*
* To avoid microcode/branching etc. all instructions are mapped out to the the
* same set of registers, however, depending on the instruction, some of the
* fields might be zero or full of garbage. You will need to make sure
* you know what instruction you are using before interpreting these
* registers.
*
* Future work: it might be a good idea in the future to type this so
* only the relevant registers are available for each instruction type.

*/
#[derive(Debug, Default, PartialEq, Eq)]
pub struct DecodedInstruction {
    // Raw Instruction Decode
    pub ins: Instruction,
    pub des: u8,
    pub sr_a: u8,
    pub sr_b: u8,
    pub con: ConditionFlags,
    pub adr: u8,
    pub sr_src: StatusRegisterUpdateSource,
    // Inferred
    pub ad_l: u8,
    pub ad_h: u8,
    pub addr_inc: i8,
    pub des_ad_l: u8,
    pub des_ad_h: u8,
    pub sr_shift: u16,
    // Dereferenced
    pub sr_a_: u16,
    pub sr_b_: u16,
    pub ad_l_: u16,
    pub ad_h_: u16,
    pub con_: bool,
    pub sr: u16,
    pub npc_l_: u16,
    pub npc_h_: u16,
}

#[derive(Debug, Default)]
pub struct IntermediateRegisters {
    pub alu_output: u16,
    pub alu_status_register: u16,
    pub lmd: u16,
}

pub trait StageExecutor {
    fn execute(
        decoded_instruction: &DecodedInstruction,
        registers: &mut Registers,
        intermediate_registers: &mut IntermediateRegisters,
        // TODO: Only the memory access stage needs this. Maybe there is a clever way to only provide each stage what they need?
        mem: &MemoryPeripheral,
    );
}

pub enum ExecutionStage {
    // TODO: Work out a cleaner way to specify the data in each stage
    Execution,
    MemoryAccessAndBranchCompletion,
    WriteBack,
}
