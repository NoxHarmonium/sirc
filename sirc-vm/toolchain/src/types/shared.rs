use crate::types::data::{DataToken, EquToken};
use crate::types::instruction::InstructionToken;
use serde::Serialize;

pub const REF_TOKEN_OFFSET_SUFFIX: &'static str = ".r";
pub const REF_TOKEN_LOWER_WORD_SUFFIX: &'static str = ".l";
pub const REF_TOKEN_UPPER_WORD_SUFFIX: &'static str = ".u";

#[derive(Debug, Serialize)]
pub enum NumberType {
    Hex,
    Decimal,
}

#[derive(Debug, Serialize)]
pub struct NumberToken {
    pub value: u32,
    pub number_type: NumberType,
}

#[derive(Debug, Serialize)]
pub struct LabelToken {
    pub name: String,
}

#[derive(Debug, Serialize)]
pub struct OriginToken {
    pub offset: u32,
}

#[derive(Debug)]
pub enum Token {
    Comment(String),
    Label(LabelToken),
    Instruction(InstructionToken),
    Origin(OriginToken),
    Data(DataToken),
    Equ(EquToken),
}
