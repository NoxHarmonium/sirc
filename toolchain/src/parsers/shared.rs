use nom::branch::alt;
use nom::bytes::complete::{is_a, tag};
use nom::character::complete::{alphanumeric1, digit1, multispace0, one_of, space0};
use nom::combinator::{map, map_res, opt, recognize};
use nom::error::{ErrorKind, ParseError};
use nom::sequence::{pair, terminated, tuple};
use nom::Err;
use nom::IResult;

use crate::types::object::RefType;

use super::instruction::RefToken;

pub fn lexeme<'a, F: 'a, O, E: ParseError<&'a str>>(
    inner: F,
) -> impl FnMut(&'a str) -> IResult<&'a str, O, E>
where
    F: FnMut(&'a str) -> IResult<&'a str, O, E>,
{
    terminated(inner, multispace0)
}

fn parse_hex_(i: &str) -> IResult<&str, u16> {
    let (i, _) = tag("0x")(i)?;
    let (i, raw_digits) = is_a(&b"0123456789abcdefABCDEF"[..])(i)?;
    let hex_parse_result = u16::from_str_radix(raw_digits, 16);
    match hex_parse_result {
        Ok(hex_value) => Ok((i, hex_value)),
        Err(_) => Err(Err::Error(nom::error::Error {
            input: i,
            code: ErrorKind::Fail,
        })),
    }
}

fn parse_dec_(i: &str) -> IResult<&str, u16> {
    map_res(
        tuple((opt(one_of("+-")), recognize(digit1))),
        |(sign, number_string)| {
            if sign.is_some() {
                // Signed numbers represented in parser as unsigned for simplicity
                str::parse::<i16>(number_string).map(|signed| signed as u16)
            } else {
                str::parse::<u16>(number_string)
            }
        },
    )(i)
}

fn parse_number_(i: &str) -> IResult<&str, u16> {
    alt((parse_hex, parse_dec))(i)
}

fn parse_comma_sep_(i: &str) -> IResult<&str, ()> {
    let (i, (_, _)) = pair(tag(","), space0)(i)?;
    Ok((i, ()))
}

fn parse_range_sep_(i: &str) -> IResult<&str, ()> {
    let (i, (_, _)) = pair(tag("->"), space0)(i)?;
    Ok((i, ()))
}

pub fn parse_label_(i: &str) -> IResult<&str, &str> {
    let (i, _) = tag(":")(i)?;
    alphanumeric1(i)
}

pub fn parse_symbol_reference_postamble_(i: &str) -> IResult<&str, Option<RefType>> {
    let (i, dot_parsed) = opt(tag("."))(i)?;

    match dot_parsed {
        Some(_) => map(
            alt((tag(".r"), tag(".u"), tag(".l"))),
            |parsed_value| match parsed_value {
                ".r" => Some(RefType::Offset),
                ".u" => Some(RefType::UpperByte),
                ".l" => Some(RefType::LowerByte),
                // TODO: Proper error handling again
                _ => panic!("Unknown postamble {}", parsed_value),
            },
        )(i),
        None => Ok((i, None)),
    }
}

pub fn parse_symbol_reference_(i: &str) -> IResult<&str, RefToken> {
    let (i, _) = tag("@")(i)?;
    let (i, name) = alphanumeric1(i)?;
    let (i, optional_preamble) = parse_symbol_reference_postamble_(i)?;

    Ok((
        i,
        RefToken {
            name: String::from(name),
            ref_type: optional_preamble.unwrap_or(RefType::Implied),
        },
    ))
}

pub fn parse_hex(i: &str) -> IResult<&str, u16> {
    lexeme(parse_hex_)(i)
}

pub fn parse_dec(i: &str) -> IResult<&str, u16> {
    lexeme(parse_dec_)(i)
}

pub fn parse_number(i: &str) -> IResult<&str, u16> {
    lexeme(parse_number_)(i)
}

pub fn parse_comma_sep(i: &str) -> IResult<&str, ()> {
    lexeme(parse_comma_sep_)(i)
}

pub fn parse_range_sep(i: &str) -> IResult<&str, ()> {
    lexeme(parse_range_sep_)(i)
}

pub fn parse_label(i: &str) -> IResult<&str, &str> {
    lexeme(parse_label_)(i)
}

pub fn parse_symbol_reference(i: &str) -> IResult<&str, RefToken> {
    lexeme(parse_symbol_reference_)(i)
}
