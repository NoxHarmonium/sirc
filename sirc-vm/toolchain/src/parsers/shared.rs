use nom::branch::alt;
use nom::bytes::complete::{is_a, is_not};
use nom::character::complete::{char, digit1, multispace0, one_of, space0};
use nom::combinator::{cut, eof, map, map_res, opt, recognize};
use nom::error::{ErrorKind, ParseError};
use nom::sequence::{pair, preceded, terminated, tuple};
use nom::{AsChar, IResult};
use nom::{Err, InputTakeAtPosition, Parser};
use nom_supreme::error::ErrorTree;
use nom_supreme::error::{BaseErrorKind, Expectation};
use nom_supreme::multi::collect_separated_terminated;
use nom_supreme::tag::complete::tag;
use nom_supreme::ParserExt;

use super::instruction::{parse_instruction_token, ShiftDefinitionData};
use crate::parsers::data::{parse_data_token, parse_equ_token};
use crate::types::data::RefToken;
use crate::types::object::RefType;
use crate::types::shared::{
    LabelToken, NumberToken, NumberType, OriginToken, Token, REF_TOKEN_LOWER_WORD_SUFFIX,
    REF_TOKEN_OFFSET_SUFFIX, REF_TOKEN_UPPER_WORD_SUFFIX,
};
use peripheral_cpu::coprocessors::processing_unit::definitions::{ShiftOperand, ShiftType};

pub type AsmResult<'a, 'b, O> = IResult<&'a str, O, ErrorTree<&'b str>>;

#[allow(clippy::type_repetition_in_bounds)]
pub fn lexeme<'a, F, O, E: ParseError<&'a str>>(
    inner: F,
) -> impl FnMut(&'a str) -> IResult<&'a str, O, E>
where
    F: FnMut(&'a str) -> IResult<&'a str, O, E> + 'a,
{
    terminated(inner, multispace0)
}

fn parse_label_name_(i: &str) -> AsmResult<&str> {
    i.split_at_position1_complete(
        |item| !item.is_alphanum() && item != '_',
        ErrorKind::AlphaNumeric,
    )
}

fn parse_hex_<T: num_traits::Num>(i: &str) -> AsmResult<T> {
    let (i, _) = tag("0x")(i)?;
    let (i, raw_digits) = is_a(&b"0123456789abcdefABCDEF"[..])(i)?;
    let hex_parse_result = T::from_str_radix(raw_digits, 16);
    hex_parse_result.map_or_else(
        |_| {
            Err(Err::Error(ErrorTree::Base {
                location: i,
                kind: BaseErrorKind::Expected(Expectation::HexDigit),
            }))
        },
        |hex_value| Ok((i, hex_value)),
    )
}

#[allow(clippy::cast_sign_loss)]
fn parse_dec_(i: &str) -> AsmResult<u32> {
    map_res(
        tuple((opt(one_of("+-")), recognize(digit1))),
        |(sign, number_string)| {
            sign.map_or_else(
                || str::parse::<u32>(number_string),
                |sign_value| {
                    // TODO: Fix strangeness in number parsing
                    // category=Toolchain
                    // Re-concatenating the original string seems bad
                    // We should probably just get the original value or something
                    let full_number = format!("{sign_value}{number_string}");
                    // Signed numbers represented in parser as unsigned for simplicity
                    str::parse::<i32>(full_number.as_str()).map(|signed| signed as u32)
                },
            )
        },
    )(i)
}

fn parse_number_(i: &str) -> AsmResult<NumberToken> {
    preceded(char('#'), alt((parse_hex, parse_dec)))(i)
}

fn parse_placeholder_(i: &str) -> AsmResult<String> {
    map(preceded(char('$'), parse_label_name_), ToOwned::to_owned)(i)
}

fn parse_comma_sep_(i: &str) -> AsmResult<()> {
    let (i, (_, _)) = pair(char(','), space0)(i)?;
    Ok((i, ()))
}

pub fn parse_label_(i: &str) -> AsmResult<&str> {
    preceded(char(':'), cut(parse_label_name_))(i)
}

pub fn parse_origin_(i: &str) -> AsmResult<NumberToken> {
    let (i, (_, value)) = tuple((lexeme(tag(".ORG")), alt((parse_hex, parse_dec))))(i)?;

    Ok((i, value))
}

pub fn parse_symbol_reference_postamble_(i: &str) -> AsmResult<Option<RefType>> {
    // TODO: Investigate why label reference suffixes aren't parsing
    // category=Toolchain
    // E.g. ".DW @some_label.l" is not parsing
    let (i, dot_parsed) = opt(tag("."))(i)?;

    match dot_parsed {
        Some(_) => map(
            alt((
                tag(REF_TOKEN_OFFSET_SUFFIX),
                tag(REF_TOKEN_UPPER_WORD_SUFFIX),
                tag(REF_TOKEN_LOWER_WORD_SUFFIX),
            )),
            |parsed_value| match parsed_value {
                REF_TOKEN_OFFSET_SUFFIX => Some(RefType::Offset),
                REF_TOKEN_UPPER_WORD_SUFFIX => Some(RefType::UpperWord),
                REF_TOKEN_LOWER_WORD_SUFFIX => Some(RefType::LowerWord),
                // TODO: Error handling in shared parsers
                // category=Refactoring
                // Return of the panic!
                _ => panic!("Unknown postamble {parsed_value}"),
            },
        )(i),
        None => Ok((i, None)),
    }
}

pub fn parse_symbol_reference_(i: &str) -> AsmResult<RefToken> {
    let (i, name) = preceded(char('@'), parse_label_name_)(i)?;

    let (i, optional_postamble) = parse_symbol_reference_postamble_(i)?;

    Ok((
        i,
        RefToken {
            name: String::from(name),
            ref_type: optional_postamble.unwrap_or(RefType::Implied),
        },
    ))
}

pub fn parse_hex(i: &str) -> AsmResult<NumberToken> {
    map(lexeme(parse_hex_), |value| NumberToken {
        value,
        number_type: NumberType::Hex,
    })(i)
}

pub fn parse_dec(i: &str) -> AsmResult<NumberToken> {
    map(lexeme(parse_dec_), |value| NumberToken {
        value,
        number_type: NumberType::Decimal,
    })(i)
}

pub fn parse_number(i: &str) -> AsmResult<NumberToken> {
    lexeme(parse_number_)(i)
}

pub fn parse_placeholder(i: &str) -> AsmResult<String> {
    lexeme(parse_placeholder_)(i)
}

pub fn parse_comma_sep(i: &str) -> AsmResult<()> {
    lexeme(parse_comma_sep_)(i)
}

pub fn parse_label(i: &str) -> AsmResult<&str> {
    lexeme(parse_label_)(i)
}

pub fn parse_origin(i: &str) -> AsmResult<u32> {
    map(lexeme(parse_origin_), |token| token.value)(i)
}

pub fn parse_symbol_reference(i: &str) -> AsmResult<RefToken> {
    lexeme(parse_symbol_reference_)(i)
}

///
/// Takes the parser representation of a shift and splits it out into the components
/// that get encoded into instructions.
///
/// In this case of instructions, the "`shift_count`" is either the index of the register,
/// or a constant depending on the value of `ShiftOperand`.
///
// TODO: Consider storing shift as an enum
// category=Refactoring
// Not sure if still valid, needs looking in to
// Should shift be stored as an enum in the instruction structs? then it could be reused by the parser and avoid this function
pub fn split_shift_definition_data(
    shift_definition_data: &ShiftDefinitionData,
) -> (ShiftOperand, ShiftType, u8) {
    match shift_definition_data {
        // TODO: More clean up in shared parser code
        // category=Refactoring
        // Probably can avoid this wrapping/unwrapping by using one type or something?
        ShiftDefinitionData::Immediate(shift_type, shift_count) => {
            (ShiftOperand::Immediate, *shift_type, *shift_count)
        }
        ShiftDefinitionData::Register(shift_type, shift_count) => (
            ShiftOperand::Register,
            *shift_type,
            shift_count.to_register_index(),
        ),
    }
}

fn parse_label_token(i: &str) -> AsmResult<Token> {
    let (i, name) = parse_label(i)?;
    Ok((
        i,
        Token::Label(LabelToken {
            name: String::from(name),
        }),
    ))
}

fn parse_origin_token(i: &str) -> AsmResult<Token> {
    let (i, offset) = parse_origin(i)?;
    Ok((i, Token::Origin(OriginToken { offset })))
}

fn parse_comment_(i: &str) -> AsmResult<Token> {
    // TODO: Should there be a more flexible parser for eol?
    // category=Toolchain
    // TODO: Comments with nothing after the semicolon currently fail
    // category=Toolchain
    map(pair(char(';'), cut(is_not("\n\r"))), |(_, text)| {
        Token::Comment(String::from(text))
    })(i)
}

fn parse_comment(i: &str) -> AsmResult<Token> {
    lexeme(parse_comment_)(i)
}

// Addresses are replaced with indexes to object table and resolved by linker
pub fn parse_tokens(i: &str) -> AsmResult<Vec<Token>> {
    let mut parser = collect_separated_terminated(
        alt((
            parse_comment.context("comment"),
            parse_instruction_token.context("instruction"),
            parse_label_token.context("label"),
            parse_origin_token.context("origin"),
            parse_data_token.context("data directive"),
            parse_equ_token.context("equ directive"),
        )),
        multispace0,
        eof,
    );

    // Consume any extra space at the start
    let (i, _) = multispace0(i)?;
    parser.parse(i)
}
