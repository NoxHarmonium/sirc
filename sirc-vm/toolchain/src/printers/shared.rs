use crate::printers::data::{print_data_token, print_equ_token};
use crate::types::data::RefToken;
use crate::types::object::RefType;
use crate::types::shared::{
    LabelToken, NumberToken, NumberType, Token, REF_TOKEN_LOWER_WORD_SUFFIX,
    REF_TOKEN_OFFSET_SUFFIX, REF_TOKEN_UPPER_WORD_SUFFIX,
};
use itertools::Itertools;

/// Prints the AST representation of a `NumberToken` to a string
///
///```
/// use toolchain::printers::shared::print_number_token;
/// use toolchain::types::shared::{NumberToken, NumberType};
/// let printed = print_number_token(&NumberToken {
///    value: 0xCAFE,
///    number_type: NumberType::Decimal
/// });
/// assert_eq!(String::from("#51966"), printed);
/// ```
pub fn print_number_token(number_token: &NumberToken) -> String {
    match number_token.number_type {
        NumberType::Hex => {
            if number_token.value > 0xFFFF {
                format!("#0x{:08X}", number_token.value)
            } else {
                format!("#0x{:04X}", number_token.value)
            }
        }
        NumberType::Decimal => format!("#{}", number_token.value),
    }
}

/// Prints the AST representation of a `RefToken` to a string
///
///```
/// use toolchain::printers::data::print_equ_token;
/// use toolchain::printers::shared::print_ref_token;
/// use toolchain::types::data::{EquToken, RefToken};
/// use toolchain::types::object::RefType;
/// use toolchain::types::shared::{NumberToken, NumberType};
/// let printed = print_ref_token(&RefToken {
///    name: String::from("some_other_symbol"),
///    ref_type: RefType::UpperWord
/// });
/// assert_eq!(String::from("@some_other_symbol.u"), printed);
/// ```
pub fn print_ref_token(ref_token: &RefToken) -> String {
    let optional_postamble = match ref_token.ref_type {
        RefType::Offset => String::from(REF_TOKEN_OFFSET_SUFFIX),
        RefType::LowerWord => String::from(REF_TOKEN_LOWER_WORD_SUFFIX),
        RefType::UpperWord => String::from(REF_TOKEN_UPPER_WORD_SUFFIX),
        _ => String::from(""),
    };

    return format!("@{}{}", ref_token.name, optional_postamble);
}

/// Prints an AST `Token` to a string
///
/// This is the generic function that should work over any token
///
///```
/// use toolchain::printers::shared::print_token;
/// use toolchain::types::data::{EquToken, RefToken};
/// use toolchain::types::object::RefType;
/// use toolchain::types::shared::{NumberToken, NumberType, Token};
/// let printed = print_token(&Token::Equ(
///    EquToken {
///        placeholder_name: String::from("BAR"),
///        number_token: NumberToken {
///            number_type: NumberType::Decimal,
///            value: 1234
///        }
///    }
/// ));
/// assert_eq!(String::from(".EQU $BAR #1234"), printed);
/// ```
pub fn print_token(token: &Token) -> String {
    match token {
        Token::Comment(text) => format!(";{text}"),
        Token::Label(LabelToken { name }) => format!(":{name}"),
        Token::Instruction(_) => todo!(),
        Token::Origin(_) => todo!(),
        Token::Data(data_token) => print_data_token(data_token),
        Token::Equ(equ_token) => print_equ_token(equ_token),
    }
}

/// Prints an AST tree to a string
///
///```
/// use toolchain::printers::shared::print_tokens;
/// use toolchain::types::data::{EquToken, RefToken};
/// use toolchain::types::object::RefType;
/// use toolchain::types::shared::{NumberToken, NumberType, Token};
/// let printed = print_tokens(vec![
///    Token::Equ(
///        EquToken {
///            placeholder_name: String::from("FOO"),
///            number_token: NumberToken {
///                number_type: NumberType::Decimal,
///                value: 1234
///            }
///        }
///    ),
///    Token::Equ(
///        EquToken {
///            placeholder_name: String::from("BAR"),
///            number_token: NumberToken {
///                number_type: NumberType::Hex,
///                value: 0xCAFE
///            }
///        }
///    ),
/// ]);
/// assert_eq!(r".EQU $FOO #1234
///.EQU $BAR #0xCAFE
///", printed);
/// ```
pub fn print_tokens(tokens: Vec<Token>) -> String {
    // TODO: Don't discard whitespace when printing
    // category=Toolchain
    // If you parse a file and then print it out again, you lose all the vertical whitespace
    tokens.iter().map(|token| print_token(token)).join("\n") + "\n"
}
