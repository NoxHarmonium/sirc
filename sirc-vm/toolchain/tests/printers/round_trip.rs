use insta::assert_snapshot;
use nom_supreme::{
    error::ErrorTree,
    final_parser::{final_parser, Location},
};
use toolchain::parsers::shared::parse_tokens;
use toolchain::printers::shared::print_tokens;
use toolchain::types::shared::Token;

static PARSER_INPUT: &str = r"
; EQU Tests
.EQU $SOME_PLACEHOLDER          #0xCAFE
.EQU $SOME_OTHER_PLACEHOLDER    #1234

:some_label

; Data Tests
.DB #0xFF
.DW #0xFFFF
.DQ #0xFFFFFFFF
.DB #255
.DB #65535
.DQ #4294967295
.DB @some_label
.DW @some_label
.DQ @some_label
";

#[test]
fn test_printing_round_trip() {
    let tokens = match final_parser::<&str, Vec<Token>, ErrorTree<&str>, ErrorTree<Location>>(
        parse_tokens,
    )(PARSER_INPUT)
    {
        Ok(tokens) => tokens,
        Err(error) => panic!("Error parsing file:\n{error}"),
    };

    let printed = print_tokens(&tokens);

    assert_snapshot!(printed);
}
