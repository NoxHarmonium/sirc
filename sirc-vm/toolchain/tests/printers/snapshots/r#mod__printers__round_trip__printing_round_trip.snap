---
source: toolchain/tests/printers/round_trip.rs
expression: printed
---
; EQU Tests
.EQU $SOME_PLACEHOLDER #0xCAFE
.EQU $SOME_OTHER_PLACEHOLDER #1234
:some_label
; Data Tests
.DB #0x00FF
.DW #0xFFFF
.DQ #0xFFFFFFFF
.DB #255
.DB #65535
.DQ #4294967295
.DB @some_label
.DW @some_label
.DQ @some_label
